com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java

```diff
@@ -107,7 +107,7 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
 	 * @param action the operation which would like to use the initialized VIATRA engine - 
 	 *               the action should not throw an exception during execution
 	 */
-	public void executeActionOnEngine(Consumer<AdvancedViatraQueryEngine> action) {
+	public synchronized void executeActionOnEngine(Consumer<AdvancedViatraQueryEngine> action) {
```

Comment:
This is sadly not enough as the `isInitialized` variable is checked in the getInitializedEngine() method in a way that there is still a race condition when it is accessed from multiple threads:

 1. The getInitializedEngine method gets called from a thread, the engine is not initialized yet so gets a lock and enters the `initializeEngine` method.
 2. Before that method finishes, the getInitializedEngine is called again, it checks whether the engine is initialized yet, and if not, tries to enter the initializeEngine where it will be waiting until the other thread stops.

As far as I can see, the second initializeEngine call would not cause any issues as it does not reinitialize the engine if not necessary and the initializationActions would be emptied. However, I'd avoid this issue by synchronizing **all** the engine initialization methods **and** do the same for the `requireQueries` methods as well.